tests/run-with-compiler-custom-args/staged-streams_1.scala

```diff
@@ -0,0 +1,697 @@
+      stream match {
+        case linear: Linear[A] => {
+          val enhancedProducer: Producer[(Var[Int], A)] = addCounter[A](n, linear.producer)
+          val enhancedStream: Linear[(Var[Int], A)] = Linear(enhancedProducer)
+
+          // Map an enhanced stream to a stream that produces the elements. Before
+          // invoking the continuation for the element, "use" the variable accordingly.
+          mapRaw[(Var[Int], A), A]((t: (Var[Int], A)) => k => '{
+            ~t._1.update('{~t._1.get - 1})
+            ~k(t._2)
+          }, enhancedStream)
+        }
+        case nested: Nested[A, bt] => {
+          val enhancedProducer: Producer[(Var[Int], bt)] = addCounter[bt](n, nested.producer)
+
+          Nested(enhancedProducer, (t: (Var[Int], bt)) => {
+            // Before invoking the continuation for the element, "use" the variable accordingly.
+            // In contrast to the linear case, the variable is initialized in the originating stream.
+            mapRaw[A, A]((el => k => '{
+              ~t._1.update('{~t._1.get - 1})
+              ~k(el)
+            }), addTerminationCondition(b => '{ ~t._1.get > 0 && ~b}, nested.nestedf(t._2)))
+          })
+        }
+      }
+     }
+
+    /** A stream containing the first `n` elements of this stream. */
+    def take(n: Expr[Int]): Stream[A] = Stream(takeRaw[Expr[A]](n, stream))
+
+    private def zipRaw[A: Type, B: Type](stream1: StagedStream[Expr[A]], stream2: StagedStream[B]): StagedStream[(Expr[A], B)] = {
+      (stream1, stream2) match {
+
+        case (Linear(producer1), Linear(producer2)) =>
+          Linear(zip_producer(producer1, producer2))
+
+        case (Linear(producer1), Nested(producer2, nestf2)) =>
+          pushLinear[Expr[A], _, B](producer1, producer2, nestf2)
+
+        case (Nested(producer1, nestf1), Linear(producer2)) =>
+          mapRaw[(B, Expr[A]), (Expr[A], B)]((t => k => '{ ~k((t._2, t._1)) }), pushLinear[B, _, Expr[A]](producer2, producer1, nestf1))
+
+        case (Nested(producer1, nestf1), Nested(producer2, nestf2)) =>
+          zipRaw[A, B](Linear(makeLinear(stream1)), stream2)
+      }
+    }
+
+    /** Make a stream linear
+      *
+      * Performs reification of the `stream`. It converts it to a function that will, when called, produce the current element
+      * and advance the stream -- or report the end-of-stream.
+      * The reified stream is an imperative *non-recursive* function, called `adv`, of `Unit => Unit` type. Nested streams are
+      * also handled.
+      *
+      * @example {{{
+      *
+      *    Stream.of(1,2,3).flatMap(d => ...)
+      *          .zip(Stream.of(1,2,3).flatMap(d => ...))
+      *          .map{ case (a, b) => a + b }
+      *          .fold(0)((a, b) => a + b)
+      * }}}
+      *
+      * -->
+      *
+      * {{{
+      *           /* initialization for stream 1 */
+      *
+      *           var curr = null.asInstanceOf[Int]  // keeps each element from reified stream
```

Comment:
This should probably be `0` instead of `null`. Also, it seems that `curr` is never updated.

Reply:
👍 

I was "updating" it in comments. Fixing it!